Add fallback wrapper for profiler

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37194 Add fallback wrapper for profiler**

Benchmark
```
import torch, time

NITER = 10000

inst = torch.classes._TorchScriptTesting._PickleTester([1, 3, 3, 7])

with torch.autograd.profiler.profile() as prof:
    s = time.time()
    for _ in range(NITER):
        torch.ops._TorchScriptTesting.take_an_instance(inst)
    e = time.time()

print('time per iter (ms)', (e - s) / 1000)
```

Before this change
```
time per iter (ms) 0.00014818763732910156
```

After this change
```
time per iter (ms) 0.00014342880249023436
```

Differential Revision: [D21217886](https://our.internmc.facebook.com/intern/diff/D21217886)

Commit:
Add fallback wrapper for profiler

[ghstack-poisoned]

jamesr66a commented on torch/csrc/autograd/profiler.cpp:
```diff
@@ -346,3 +346,15 @@ void RecordProfile::processEvents(const std::vector<Event*>& events) {
+  .registerFallback(
+    c10::DispatchKey::Profiler,
+    c10::KernelFunction::makeFromBoxedFunction<&profile_wrapper>(),
+    "");
```
@ezyang I copied in the code from a very old revision of my other PR. Let me know if we should modernize these API calls or anything

Also, should we add `callBoxedRedispatch`?

ezyang replied:
Yeah let's modernize. Also, let's register makeFallthrough function (rather than doing it manually); it will be faster

Commit:
Update on "Add fallback wrapper for profiler"


Differential Revision: [D21217886](https://our.internmc.facebook.com/intern/diff/D21217886)

[ghstack-poisoned]

Commit:
Update on "Add fallback wrapper for profiler"


Benchmark
```
import torch, time

NITER = 10000

inst = torch.classes._TorchScriptTesting._PickleTester([1, 3, 3, 7])

with torch.autograd.profiler.profile() as prof:
    s = time.time()
    for _ in range(NITER):
        torch.ops._TorchScriptTesting.take_an_instance(inst)
    e = time.time()

print('time per iter (ms)', (e - s) / 1000)
```

Before this change
```
time per iter (ms) 0.00014818763732910156
```

After this change
```
time per iter (ms) 0.00014342880249023436
```

Differential Revision: [D21217886](https://our.internmc.facebook.com/intern/diff/D21217886)

[ghstack-poisoned]
